pkg/containerdutil/snapshotter.go

```diff
@@ -0,0 +1,63 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package containerdutil
+
+import (
+	"context"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/snapshots"
+)
+
+type Snapshotter = snapshots.Snapshotter
+
+// SnapshotService should be called to get a new caching snapshotter
+func SnapshotService(client *containerd.Client, snapshotterName string) Snapshotter {
+	return &snapshotterWithCache{
+		client.SnapshotService(snapshotterName),
+		map[string]snapshots.Info{},
+		map[string]snapshots.Usage{},
+	}
+}
+
+type snapshotterWithCache struct {
+	snapshots.Snapshotter
+	statCache  map[string]snapshots.Info
+	usageCache map[string]snapshots.Usage
```

Comment:
They can change on every read (e.g. Info.Updated). I don't think they can be cached.

Reply:
Thanks a lot for the review @ktock.

This is not meant to be used for *successive* operations.
The purpose of this PR and new objects overall is *solely* for use for single, atomic methods retrieving a bunch of objects at once (`nerdctl ps`, `merdctl images`), to avoid useless, successive reads of the same objects (as discussed in the linked issues and initial comment).

I did put a note here about that: https://github.com/containerd/nerdctl/pull/3029/files/8eba956203463c6e2ccce9afea8dae31a9149908#diff-c9a808f7f845733683410c3156f957135eba140a8783144049c9f2aea14251ccR17

Specifically, when multiple different tags point to the same image, it does not make sense to re-read `usage` once per tag (even if it would have changed between the time we read for the first tag)- and if we would, it does not make sense either to return different size information...

Hope that clarifies.